taskana-adapter-camunda-outbox-rest/src/main/java/pro/taskana/adapter/camunda/outbox/rest/service/CamundaTaskEventsService.java

```diff
@@ -330,24 +388,59 @@ public CamundaTaskEvent getEvent(int id) throws CamundaTaskEventNotFoundExceptio
     throw new CamundaTaskEventNotFoundException("camunda task event not found");
   }
 
-  public List<CamundaTaskEvent> getAllEvents() {
+  public List<CamundaTaskEvent> getAllEvents(Duration lockDuration) {
 
     List<CamundaTaskEvent> camundaTaskEvents = new ArrayList<>();
-
-    String sql = String.format(SQL_GET_ALL_EVENTS, OUTBOX_SCHEMA);
-
+    String sqlStatement = lockDuration == null
+        ? SQL_GET_ALL_EVENTS : SQL_GET_ALL_AVAILABLE_EVENTS;
+    String sql = String.format(sqlStatement, OUTBOX_SCHEMA);
+    List<Integer> ids = null;
     try (Connection connection = getConnection();
         PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
-
-      ResultSet camundaTaskEventResultSet = preparedStatement.executeQuery();
-      camundaTaskEvents = getCamundaTaskEvents(camundaTaskEventResultSet);
-
+      try {
+        if (lockDuration != null) {
+          preparedStatement.setTimestamp(1, Timestamp.from(Instant.now()));
+        }
+        ResultSet camundaTaskEventResultSet = preparedStatement.executeQuery();
+        camundaTaskEvents = getCamundaTaskEvents(camundaTaskEventResultSet);
+        ids = camundaTaskEvents.stream().map(CamundaTaskEvent::getId)
+            .collect(Collectors.toList());
+        lockEvents(ids, lockDuration, connection);
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug(
+              "Events locked: {}",
+              ids.stream().map(Object::toString).collect(Collectors.joining(";\n")));
+        }
+      } catch (Exception e) {
+        LOGGER.warn("Caught Exception while trying to retrieve all events from the outbox",
+            e);
+        if (ids != null) {
+          try {
+            unlockEvents(ids, connection);
+            if (LOGGER.isDebugEnabled()) {
+              LOGGER.debug(
+                  "Events unlocked: {}",
+                  ids.stream().map(Object::toString).collect(Collectors.joining(";\n")));
+            }
+          } catch (Exception ex) {
+            LOGGER.error("Failed to unlock events", ex);
+          }
+        }
+      }
     } catch (Exception e) {
       LOGGER.warn("Caught Exception while trying to retrieve all events from the outbox", e);
     }
     return camundaTaskEvents;
   }
 
+  public void unlockEventsForId(Integer eventId) {
```

Comment:
this always unlocks exactely one event, right? If yes, what do you think about naming it unlockEventForId?

Reply:
Yes, I will rename it